### PR TITLE
[RISCV][GISel] Legalize G_{U|S}{ADD|SUB}SAT

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for G_UADDSAT/G_SADDSAT/G_USUBSAT/G_SSUBSAT by lowering it into add/sub with selects. When zbb is available, min/max/minu/maxu will be selected.

I would like to leave the vector support (vs{add|sub}u?) as a todo until I fix all crashes in [my out-of-tree tests](https://github.com/dtcxzyw/llvm-codegen-benchmark).

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

**Comment by tschuett** [comment]
Legalize or legalize and select? There are no assembler files.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-sat-rv64.mir [hunk 1]
```diff
@@ -0,0 +1,358 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=riscv64 -run-pass=legalizer %s -o - \
+# RUN: | FileCheck %s --check-prefixes=CHECK,RV64I
+# RUN: llc -mtriple=riscv64 -mattr=+zbb -run-pass=legalizer %s -o - \
```
Comment (tschuett) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-sat-rv64.mir:4]: +zbb has no effect?

Comment (tschuett) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-sat-rv64.mir:4]: ignore that.

**Comment by dtcxzyw** [comment]
> Legalize or legalize and select? There are no assembler files.

See the check line `CHECK,RV32ZBB`. TBH, I think the MIR tests are enough. Should I also add some codegen tests?

**Comment by arsenm** [comment]
> > Legalize or legalize and select? There are no assembler files.
> 
> See the check line `CHECK,RV32ZBB`. TBH, I think the MIR tests are enough. Should I also add some codegen tests?

End to end codegen tests should generally be preferred

**Comment by tschuett** [comment]
See the check line CHECK,RV32ZBB. TBH, I think the MIR tests are enough. Should I also add some codegen tests?

Up to you, but there are no assembler RISC-V files. I would change the title to Legalize.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -361,6 +361,10 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   getActionDefinitionsBuilder(G_DYN_STACKALLOC).lower();
 
+  // TODO: Use Vector Single-Width Saturating Instructions for vector types.
+  getActionDefinitionsBuilder({G_UADDSAT, G_SADDSAT, G_USUBSAT, G_SSUBSAT})
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:365]: Maybe put this closer to the legalization rules fro ADD and ADDO?

**Pull request merged**